README.md
<a href="https://www.buymeacoffee.com/sBGXj7Pl4" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: 41px !important;width: 174px !important;box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;" ></a>

# Buy me a coffee widget

Did you ever needed a widget for buy me a coffee, well here it is.

![Buy me a coffee widget](https://github.com/md-weber/buy_me_a_coffee/blob/master/assets/readme/BuyMeACoffee.gif)

## Important note
Be very careful with using this widget. According to the Google Play Guidelines you are not 
allowed to get payments from external ressources. So a donation button could lead to a 
ban of your app!

This app should only be used in Web or in Stores that allow a donation.

For more information please read this 
* [Reddit Post](https://www.reddit.com/r/androiddev/comments/7cfisr/does_google_unpublish_apps_for_having_a_donate/).
* [Google Play Store](https://play.google.com/about/monetization-ads/)

## Getting Started

You have to import

```dart
import 'package:buy_me_a_coffee_widget/buy_me_a_coffee_widget.dart';
```

then you can use the Widget.

```dart
Container(
  width: 217.0,
  child: BuyMeACoffeeWidget(
      sponsorID: "sBGXj7Pl4",
      theme: theme,
  ),
)
```

For a more in depth example please go to the [Example Folder](./example).



CHANGELOG.md
# Changelog

## [2.0.0-nullsafety.0] - 14.12.2020

* Migrated to null-saftey
* Removed Google Fonts
* Currently Roboto is used - You have to Provide own font

## [1.0.1] - 03.11.2020

* Version update of google_fonts && flutter_svg

## [1.0.0] - 28.01.2020

* Added an important note to the readme file.

## [0.0.2] - 28.01.2020

* Added the example to the project


## [0.0.1] - 28.01.2020

* First release of the package. You can use the widget now and change the appearance with the different Themes that are provided.

example/lib/main.dart
import 'package:buy_me_a_coffee_widget/buy_me_a_coffee_widget.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  // This widget is the root of your application.
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  BuyMeACoffeeThemeData theme = OrangeTheme();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text("Buy me a coffee Widget"),
        ),
        body: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Center(
              child: Container(
                padding: EdgeInsets.symmetric(horizontal: 40),
                child: BuyMeACoffeeWidget(
                  sponsorID: "flutterexp",
                  theme: theme,
                ),
              ),
            ),
            SizedBox(height: 20),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Column(
                children: <Widget>[
                  Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: <Widget>[
                      ThemeSelector(
                        color: Color(0xFFFF813F),
                        onTap: () {
                          setState(() {
                            theme = OrangeTheme();
                          });
                        },
                      ),
                      ThemeSelector(
                        color: Color(0xFFFFDD00),
                        onTap: () {
                          setState(() {
                            theme = YellowTheme();
                          });
                        },
                      ),
                      ThemeSelector(
                        color: Color(0xFFbd5fff),
                        onTap: () {
                          setState(() {
                            theme = PurpleTheme();
                          });
                        },
                      ),
                      ThemeSelector(
                        color: Colors.black,
                        onTap: () {
                          setState(() {
                            theme = BlackTheme();
                          });
                        },
                      ),
                    ],
                  ),
                  Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: <Widget>[
                      ThemeSelector(
                        color: Colors.white,
                        onTap: () {
                          setState(() {
                            theme = WhiteTheme();
                          });
                        },
                      ),
                      ThemeSelector(
                        color: Color(0xFF79D6B5),
                        onTap: () {
                          setState(() {
                            theme = TealTheme();
                          });
                        },
                      ),
                      ThemeSelector(
                        color: Color(0xFFFF5F5F),
                        onTap: () {
                          setState(() {
                            theme = RedTheme();
                          });
                        },
                      ),
                      ThemeSelector(
                        color: Color(0xFFF471FF),
                        onTap: () {
                          setState(() {
                            theme = PinkTheme();
                          });
                        },
                      )
                    ],
                  ),
                ],
              ),
            )
          ],
        ),
      ),
    );
  }
}

class ThemeSelector extends StatelessWidget {
  final VoidCallback onTap;
  final Color color;

  const ThemeSelector({Key? key, required this.onTap, required this.color})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: GestureDetector(
        onTap: onTap,
        child: new Container(
            child: new CircleAvatar(backgroundColor: color),
            width: 64.0,
            height: 64.0,
            padding: const EdgeInsets.all(1.0),
            decoration: new BoxDecoration(
              color: Colors.black, // border color
              shape: BoxShape.circle,
            )),
      ),
    );
  }
}


lib/themes/themes.dart
import 'package:flutter/material.dart';

/// This is the interface for all Themes for BuyMeACoffee
abstract class BuyMeACoffeeThemeData {
  /// Background Color of the Button
  late Color backgroundColor;

  /// Text Color inside the button
  late Color textColor;
}

/// [OrangeTheme] is a implementation of [BuyMeACoffeeThemeData]
/// it has a orange background and a white text color
class OrangeTheme implements BuyMeACoffeeThemeData {
  Color backgroundColor = Color(0xFFFF813F);
  Color textColor = Colors.white;
}

/// [YellowTheme] is a implementation of [BuyMeACoffeeThemeData]
/// it has a yellow background and a black text color
class YellowTheme implements BuyMeACoffeeThemeData {
  Color backgroundColor = Color(0xFFFFDD00);
  Color textColor = Colors.black;
}

/// [PurpleTheme] is a implementation of [BuyMeACoffeeThemeData]
/// it has a Purple background and a white text color
class PurpleTheme implements BuyMeACoffeeThemeData {
  Color backgroundColor = Color(0xFFbd5fff);
  Color textColor = Colors.white;
}

/// [BlackTheme] is a implementation of [BuyMeACoffeeThemeData]
/// it has a black background and a white text color
class BlackTheme implements BuyMeACoffeeThemeData {
  Color backgroundColor = Colors.black;
  Color textColor = Colors.white;
}

/// [WhiteTheme] is a implementation of [BuyMeACoffeeThemeData]
/// it has a white background and a black text color
class WhiteTheme implements BuyMeACoffeeThemeData {
  Color backgroundColor = Colors.white;
  Color textColor = Colors.black;
}

/// [BlueTheme] is a implementation of [BuyMeACoffeeThemeData]
/// it has a blue background and a white text color
class BlueTheme implements BuyMeACoffeeThemeData {
  Color backgroundColor = Color(0xFF5F7FFF);
  Color textColor = Colors.white;
}

/// [TealTheme] is a implementation of [BuyMeACoffeeThemeData]
/// it has a teal background and a white text color
class TealTheme implements BuyMeACoffeeThemeData {
  Color backgroundColor = Color(0xFF79D6B5);
  Color textColor = Colors.white;
}

/// [RedTheme] is a implementation of [BuyMeACoffeeThemeData]
/// it has a lightRed background and a white text color
class RedTheme implements BuyMeACoffeeThemeData {
  Color backgroundColor = Color(0xFFFF5F5F);
  Color textColor = Colors.white;
}

/// [PinkTheme] is a implementation of [BuyMeACoffeeThemeData]
/// it has a pink background and a white text color
class PinkTheme implements BuyMeACoffeeThemeData {
  Color backgroundColor = Color(0xFFF471FF);
  Color textColor = Colors.white;
}


lib/src/buy_me_a_coffee_widget.dart
import 'package:buy_me_a_coffee_widget/themes/themes.dart';
import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';
import 'package:url_launcher/url_launcher.dart';

/// Url to the buy me a coffee website
const String buyMeACoffeeUrl = "https://www.buymeacoffee.com/";

/// The [BuyMeACoffeeWidget] is the widget. It has a tap event and will trigger as soon someone taps it.
/// It opens up a browser window in the default browser of the device and
/// navigates to the specified UserID on Buy me a coffee.
class BuyMeACoffeeWidget extends StatelessWidget {
  /// The id for the user where it should link to.
  final String sponsorID;

  /// Custom text for the widget
  final String customText;

  /// Overwrites the textStyle of the widget
  final TextStyle? textStyle;

  /// Color of the background if none is provided the [theme] background
  /// will be used or the fallback
  final Color? backgroundColor;

  /// The theme of the widget, it changes the appearence of the Button
  ///
  /// There are multiple themes provided
  /// - [OrangeTheme]
  /// - [YellowTheme]
  /// - [BlackTheme]
  /// - [BlueTheme]
  /// - [PurpleTheme]
  /// - [WhiteTheme]
  ///
  /// Find more [BuyMeACoffeeThemeData]
  final BuyMeACoffeeThemeData? theme;

  /// Constructor of the BuyMeACoffee Widget
  const BuyMeACoffeeWidget({
    Key? key,
    required this.sponsorID,
    this.customText = "Buy me a coffee",
    this.textStyle,
    this.backgroundColor,
    this.theme,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    var backgroundColor;
    var textStyle;

    if (theme == null && backgroundColor == null) {
      backgroundColor = Color(0xFFFF813F);
    } else if (backgroundColor != null) {
      backgroundColor = backgroundColor;
    } else {
      backgroundColor = theme!.backgroundColor;
    }

    return GestureDetector(
      onTap: () async {
        var urlString = buyMeACoffeeUrl + sponsorID;
        if (await canLaunch(urlString)) {
          await launch(urlString);
        } else {
          throw "BuyMeACoffeWidget - Something went wrong!";
        }
      },
      child: ConstrainedBox(
        constraints: BoxConstraints(minWidth: 217.0),
        child: Container(
          padding: EdgeInsets.symmetric(vertical: 7.0, horizontal: 10.0),
          height: 51.0,
          decoration: BoxDecoration(
            color: backgroundColor,
            border: Border.all(color: Colors.transparent, width: 1.0),
            boxShadow: [
              BoxShadow(
                color: Color.fromRGBO(190, 190, 190, 0.5),
                blurRadius: 2.0,
                offset: Offset.lerp(Offset(0, 0), Offset(1, 1), 1)!,
              )
            ],
            borderRadius: BorderRadius.all(
              Radius.circular(5.0),
            ),
          ),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              SvgPicture.network(
                "https://cdn.buymeacoffee.com/buttons/bmc-new-btn-logo.svg",
                width: 35.0,
              ),
              Padding(
                padding: const EdgeInsets.only(left: 8.0),
                child: Text(
                  customText,
                  style: textStyle == null
                      // TODO: Add Google Fonts as soon as null safety migration has been done
                      ? TextStyle(
                          fontFamily: "Cookie",
                          color:
                              theme == null ? Colors.white : theme!.textColor,
                          fontSize: 28.0,
                          letterSpacing: 0.6,
                        )
                      : textStyle,
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}
